include/seqan3/io/sequence/sequence_file_out.hpp

```diff
@@ -0,0 +1,900 @@
+    }
+    //!\}
+
+    //!\brief The options are public and its members can be set directly.
+    sequence_file_out_options options;
+
+    /*!\cond DEV
+     * \brief Expose a reference to the underlying stream object. [public, but not documented as part of the API]
+     */
+    stream_type & get_stream()
+    {
+        return stream;
+    }
+    //!\endcond
+protected:
+    //!\privatesection
+    //!\brief Path of the file that the stream operates on.
+    std::string file_name;
+
+    //!\brief The stream we are writing to.
+    stream_type stream;
+
+    //!\brief Type of the format, an std::variant over the `valid_formats`.
+    using format_type = detail::transfer_template_args_onto_t<valid_formats, std::variant>;
+    //!\brief The actual std::variant holding a pointer to the detected/selected format.
+    format_type format;
+
+    //!\brief Write record to format.
+    template <typename seq_t, typename id_t, typename qual_t, typename seq_qual_t>
+    void write_record(seq_t && seq, id_t && id, qual_t && qual, seq_qual_t && seq_qual)
+    {
+        assert(!format.valueless_by_exception());
+        std::visit([&] (sequence_file_out_format_concept & f)
+        {
+            // read new record
+            f.write(stream,
+                    options,
+                    seq,
+                    id,
+                    qual,
+                    seq_qual);
+        }, format);
+    }
+
+    //!\brief Write columns to file format, only tag-dispatch once.
+    template <input_range_concept seqs_t,
+              input_range_concept ids_t,
+              input_range_concept quals_t,
+              input_range_concept seq_quals_t>
+    void write_columns(seqs_t       && seqs,
+                       ids_t        && ids,
+                       quals_t      && quals,
+                       seq_quals_t  && seq_quals)
+    {
+        static_assert(!(detail::decays_to_ignore_v<reference_t<seqs_t>> &&
+                        detail::decays_to_ignore_v<reference_t<ids_t>> &&
+                        detail::decays_to_ignore_v<reference_t<quals_t>> &&
+                        detail::decays_to_ignore_v<reference_t<seq_quals_t>>),
+                      "At least one of the columns must not be set to std::ignore.");
+
+        // make sure that std::ignore inputs are actually range of std::ignore
+        assert(!format.valueless_by_exception());
+        std::visit([&] (sequence_file_out_format_concept & f)
+        {
+            auto seqs_it       = ranges::begin(seqs);
+            auto ids_it        = ranges::begin(ids);
+            auto quals_it      = ranges::begin(quals);
+            auto seq_quals_it  = ranges::begin(seq_quals);
+
+            auto seqs_end      = ranges::end(seqs);
+            auto ids_end       = ranges::end(ids);
+            auto quals_end     = ranges::end(quals);
+            auto seq_quals_end = ranges::end(seq_quals);
+
+            do
+            {
+                f.write(stream,
```

Comment:
can this not in one line? It looks ugly

Reply:
I think it follows the structure of the paragraphs before and after, it is immediately clear, that we have for things, that we 1) initialize, 2) pass on 3) increment 4) compare.

Sticking some of them in a single line reduces readability, but I can change it if you insist.